src/main/java/graphql/servlet/instrumentation/RequestStack.java

```diff
@@ -0,0 +1,304 @@
+            return Objects.equals(happenedStrategyCallsPerLevel.get(level), expectedStrategyCallsPerLevel.get(level));
+        }
+
+        private boolean allOnFieldCallsHappened(int level) {
+            return Objects.equals(happenedOnFieldValueCallsPerLevel.get(level), expectedStrategyCallsPerLevel.get(level));
+        }
+
+        private boolean allFetchesHappened(int level) {
+            return Objects.equals(fetchCountPerLevel.get(level), expectedFetchCountPerLevel.get(level));
+        }
+
+        private List<Integer> getDispatchedLevels() {
+            return dispatchedLevels;
+        }
+
+        @Override
+        public String toString() {
+            return "CallStack{" +
+                "expectedFetchCountPerLevel=" + expectedFetchCountPerLevel +
+                ", fetchCountPerLevel=" + fetchCountPerLevel +
+                ", expectedStrategyCallsPerLevel=" + expectedStrategyCallsPerLevel +
+                ", happenedStrategyCallsPerLevel=" + happenedStrategyCallsPerLevel +
+                ", happenedOnFieldValueCallsPerLevel=" + happenedOnFieldValueCallsPerLevel +
+                ", dispatchedLevels" + dispatchedLevels +
+                '}';
+        }
+
+        private boolean dispatchIfNotDispatchedBefore(int level) {
+            if (dispatchedLevels.contains(level)) {
+                Assert.assertShouldNeverHappen("level " + level + " already dispatched");
+                return false;
+            }
+            dispatchedLevels.add(level);
+            return true;
+        }
+
+        private void clearAndMarkCurrentLevelAsReady(int level) {
+            expectedFetchCountPerLevel.clear();
+            fetchCountPerLevel.clear();
+            expectedStrategyCallsPerLevel.clear();
+            happenedStrategyCallsPerLevel.clear();
+            happenedOnFieldValueCallsPerLevel.clear();
+            dispatchedLevels.clear();
+
+            // make sure the level is ready
+            expectedFetchCountPerLevel.put(level, 1);
+            expectedStrategyCallsPerLevel.put(level, 1);
+            happenedStrategyCallsPerLevel.put(level, 1);
+        }
+    }
+
+    /**
+     * Sets the status indicating if a specific execution is ready for dispatching.
+     * @param executionId must be an active execution
+     * @param toState if ready to dispatch
+     */
+    public void setStatus(ExecutionId executionId, boolean toState) {
+        if (!activeRequests.containsKey(executionId)) {
+            throw new IllegalStateException(
+                String.format("Can not set status for execution %s, it is not managed by this request stack", executionId));
+        }
+        status.put(executionId, toState);
+    }
+
+    /**
+     * @return if all managed executions are ready to be dispatched.
+     */
+    public boolean allReady() {
+        List<Integer> dispatchStack = activeRequests.values().stream().findFirst().map(CallStack::getDispatchedLevels).orElse(Collections.emptyList());
```

Comment:
not quite sure about the logic here, if activeRequests contains multiple elements, would findFirst() still make sense ?

Reply:
Each time a level of the query hierarchy is started one of the active stacks will get incremented above the others. Because the futures are created serially, once that one finishes resolving they will all evaluate as ready according to the first part of the if. To make sure we only dispatch once each active query has created their load calls for the current level we just need to make sure that they all evaluate to ready and also have the same dispatch level completed, so it doesn't matter which one we start with, they just all need to be equal

Reply:
I guess in this context the first part of the if means there are no fetch calls that are unfinished and the second part is that each query has been processed for this step of the hierarchy. The instrumentation removes active requests in the instrumentation call to instrument the result to prevent this from causing deeper queries from hanging waiting on a higher level that will never execute to that depth. Does that help?

Reply:
I guess I could do something like .stream().distinct().count() <= 1 and it would be clearer